controllers/update_status_test.go

```diff
@@ -41,6 +43,352 @@ import (
+			cluster = internal.CreateDefaultCluster()
+			err = setupClusterForTest(cluster)
+			Expect(err).NotTo(HaveOccurred())
+			// Define cluster's taint policy
+			cluster.Spec.AutomationOptions.Replacements.TaintReplacementOptions = []fdbv1beta2.TaintReplacementOption{
+				{
+					Key:               &taintKeyWildcard,
+					DurationInSeconds: &taintKeyWildcardDuration,
+				},
+				{
+					Key:               &taintKeyMaintenance,
+					DurationInSeconds: &taintKeyMaintenanceDuration,
+				},
+			}
+
+			pods, err = clusterReconciler.PodLifecycleManager.GetPods(context.TODO(), clusterReconciler, cluster, internal.GetSinglePodListOptions(cluster, "storage-1")...)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(len(pods)).To(Equal(1))
+
+			for _, container := range pods[0].Spec.Containers {
+				pods[0].Status.ContainerStatuses = append(pods[0].Status.ContainerStatuses, corev1.ContainerStatus{Ready: true, Name: container.Name})
+			}
+
+			adminClient, err = mock.NewMockAdminClientUncast(cluster, k8sClient)
+			Expect(err).NotTo(HaveOccurred())
+
+			configMap = &corev1.ConfigMap{}
+			err = k8sClient.Get(context.TODO(), types.NamespacedName{Namespace: cluster.Namespace, Name: cluster.Name + "-config"}, configMap)
+			Expect(err).NotTo(HaveOccurred())
+
+			pod = pods[0]
+			node = &corev1.Node{
+				ObjectMeta: metav1.ObjectMeta{Name: pod.Spec.NodeName},
+			}
+
+			databaseStatus, err := adminClient.GetStatus()
+			Expect(err).NotTo(HaveOccurred())
+			processMap = make(map[fdbv1beta2.ProcessGroupID][]fdbv1beta2.FoundationDBStatusProcessInfo)
+			for _, process := range databaseStatus.Cluster.Processes {
+				processID, ok := process.Locality["process_id"]
+				// if the processID is not set we fall back to the instanceID
+				if !ok {
+					processID = process.Locality["instance_id"]
+				}
+				processMap[fdbv1beta2.ProcessGroupID(processID)] = append(processMap[fdbv1beta2.ProcessGroupID(processID)], process)
+			}
+
+			allPods, err = clusterReconciler.PodLifecycleManager.GetPods(context.TODO(), clusterReconciler, cluster, internal.GetPodListOptions(cluster, "", "")...)
+			Expect(err).NotTo(HaveOccurred())
+
+			allPvcs = &corev1.PersistentVolumeClaimList{}
+			err = clusterReconciler.List(context.TODO(), allPvcs, internal.GetPodListOptions(cluster, "", "")...)
+			Expect(err).NotTo(HaveOccurred())
+
+			processGroupStatus = fdbv1beta2.FindProcessGroupByID(cluster.Status.ProcessGroups, internal.GetProcessGroupIDFromMeta(cluster, pod.ObjectMeta))
+			pvcMap := internal.CreatePVCMap(cluster, allPvcs)
+			pvcValue, pvcExists := pvcMap[processGroupStatus.ProcessGroupID]
+			if pvcExists {
+				pvc = &pvcValue
+			}
+			log.Info("Target processGroupStatus Info", "ProcessGroupID", processGroupStatus.ProcessGroupID,
+				"Conditions size", len(processGroupStatus.ProcessGroupConditions),
+				"Conditions", processGroupStatus.ProcessGroupConditions,
+				"AllPods size", len(allPods))
+		})
+
+		When("process group's node is tainted", func() {
```

Comment:
Could we restructure the tests below to fit this pattern:


```go
		When("the process group's node is tainted", func() {
			JustBeforeEach(func() {
				log.Info("Taint node", "Node name", pod.Name, "Node taints", node.Spec.Taints)
				Expect(k8sClient.Update(context.TODO(), node)).NotTo(HaveOccurred())

				err = validateProcessGroup(context.TODO(), clusterReconciler, cluster, pod, pvc, pod.ObjectMeta.Annotations[fdbv1beta2.LastConfigMapKey], processGroupStatus, cluster.IsTaintFeatureDisabled(), logger)
				Expect(err).NotTo(HaveOccurred())
			})

			When("the taint was just added", func() {
				BeforeEach(func() {
					node.Spec.Taints = []corev1.Taint{
						{
							Key:       taintKeyMaintenance,
							Value:     "rack_maintenance",
							Effect:    corev1.TaintEffectNoExecute,
							TimeAdded: &metav1.Time{Time: time.Now()},
						},
					}
				})

				It("should get NodeTaintDetected condition", func() {
					Expect(len(processGroupStatus.ProcessGroupConditions)).To(Equal(1))
					Expect(processGroupStatus.GetCondition(fdbv1beta2.NodeTaintDetected)).NotTo(Equal(nil))
				})
			})

			When("the taint was just added long enough to to mark the Pod for replacement", func() {
				BeforeEach(func() {
					node.Spec.Taints = []corev1.Taint{
						{
							Key:       taintKeyMaintenance,
							Value:     "rack_maintenance",
							Effect:    corev1.TaintEffectNoExecute,
							TimeAdded: &metav1.Time{Time: time.Now().Add(-time.Second * time.Duration(taintKeyMaintenanceDuration+1))},
						},
					}
				})

				It("should get both NodeTaintDetected and NodeTaintReplacing condition", func() {
					Expect(len(processGroupStatus.ProcessGroupConditions)).To(Equal(2))
					Expect(processGroupStatus.GetCondition(fdbv1beta2.NodeTaintDetected)).NotTo(Equal(nil))
					Expect(processGroupStatus.GetCondition(fdbv1beta2.NodeTaintReplacing)).NotTo(Equal(nil))
				})
			})
```

This reduce some of the setup code for the test cases and makes it easier to read and understand those test cases.

Reply:
I like it. will refactor it.